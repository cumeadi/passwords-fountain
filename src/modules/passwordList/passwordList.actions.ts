import { AppState, store } from '@/store';
import { Client } from 'faunadb';
import { callAction, mergeState } from '@/common/utils/store';
import { databaseActions } from '@/modules/database/database.actions';
import {
    selectIsClientSet,
    selectClient,
} from '@/modules/database/database.selectors';
import {
    OptionsPanelVariantName,
    optionsPanelVariantNames,
    PasswordEntityPayload,
    PasswordEntityPayloadReferable,
} from '@/modules/passwordList/passwordList.constants';
import { PasswordListState } from '@/modules/passwordList/passwordList.state';
import { overlayActions } from '@/modules/overlay/overlay.actions';

const merge = mergeState<PasswordListState>('passwordList');

export const passwordListActions = {
    switchOptionPanelVariant: (
        appState: AppState,
        optionPanelVariantName: OptionsPanelVariantName
    ): Partial<AppState> => {
        return merge({
            currentOptionPanelVariantName: optionPanelVariantName,
        });
    },
    resetSelectedAndDecryptedEntity: (): Partial<AppState> => {
        return merge({
            selectedAndDecryptedEntity: {} as PasswordEntityPayloadReferable,
        });
    },
    setSelectedAndDecryptedEntity: (
        appState: AppState,
        entity: PasswordEntityPayloadReferable
    ): Partial<AppState> => {
        return merge({
            selectedAndDecryptedEntity: entity,
        });
    },
    fetchPasswords: async (
        appState: AppState,
        masterKey: string,
        adminKey: string
    ): Promise<Partial<AppState>> => {
        callAction(overlayActions.showGlobalLoader);
        const { fetchAllPasswordEntities } = await import(
            '@/modules/database/database.service'
        );

        if (!selectIsClientSet(appState)) {
            await callAction(databaseActions.setClient, masterKey, adminKey);
        }

        const client = selectClient(store.getState()) as Client;

        try {
            const passwords = await fetchAllPasswordEntities(client);
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                overlayActions.showSnackbar,
                'snackbar.passwordsFetchedSuccessfully',
                'success'
            );
            callAction(
                passwordListActions.switchOptionPanelVariant,
                optionsPanelVariantNames.entityFormCollapsed
            );

            return merge({ passwords });
        } catch (err) {
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                overlayActions.showSnackbar,
                'snackbar.couldNotFetchPasswords',
                'error'
            );
            return merge({});
        }
    },
    addNewPassword: async (
        appState: AppState,
        newEntityPayload: PasswordEntityPayload,
        masterKey: string
    ): Promise<Partial<AppState>> => {
        callAction(overlayActions.showGlobalLoader);
        const { createPasswordEntity } = await import(
            '@/modules/database/database.service'
        );
        const { encrypt } = await import('@/modules/cipher/cipher.service');

        const client = selectClient(store.getState()) as Client;

        try {
            const encryptedPasswordEntity = encrypt(
                {
                    login: newEntityPayload.login,
                    password: newEntityPayload.password,
                },
                masterKey,
                true
            );
            await createPasswordEntity(client, {
                label: newEntityPayload.label,
                value: encryptedPasswordEntity,
            });
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                passwordListActions.switchOptionPanelVariant,
                optionsPanelVariantNames.entityFormCollapsed
            );
        } catch (err) {
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                overlayActions.showSnackbar,
                'snackbar.couldNotCreateNewPassword',
                'error'
            );
        } finally {
            return merge({});
        }
    },
    editPassword: async (
        appState: AppState,
        entityPayload: PasswordEntityPayloadReferable,
        masterKey: string
    ): Promise<Partial<AppState>> => {
        callAction(overlayActions.showGlobalLoader);
        const { updatePasswordEntity } = await import(
            '@/modules/database/database.service'
        );
        const { encrypt } = await import('@/modules/cipher/cipher.service');

        const client = selectClient(store.getState()) as Client;

        try {
            const encryptedPasswordEntity = encrypt(
                {
                    login: entityPayload.login,
                    password: entityPayload.password,
                },
                masterKey,
                true
            );
            await updatePasswordEntity(client, entityPayload.refId, {
                label: entityPayload.label,
                value: encryptedPasswordEntity,
            });
            callAction(overlayActions.hideGlobalLoader);
            callAction(passwordListActions.resetSelectedAndDecryptedEntity);
            callAction(
                passwordListActions.switchOptionPanelVariant,
                optionsPanelVariantNames.entityFormCollapsed
            );
        } catch (err) {
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                overlayActions.showSnackbar,
                'snackbar.couldNotEditPassword',
                'error'
            );
        } finally {
            return merge({});
        }
    },
    removePassword: async (
        appState: AppState,
        refId: string
    ): Promise<Partial<AppState>> => {
        callAction(overlayActions.showGlobalLoader);
        const { deletePasswordEntity } = await import(
            '@/modules/database/database.service'
        );

        const client = selectClient(store.getState()) as Client;

        try {
            await deletePasswordEntity(client, refId);
            callAction(overlayActions.hideGlobalLoader);
        } catch (err) {
            callAction(overlayActions.hideGlobalLoader);
            callAction(
                overlayActions.showSnackbar,
                'snackbar.couldNotRemovePassword',
                'error'
            );
        } finally {
            return merge({});
        }
    },
};
