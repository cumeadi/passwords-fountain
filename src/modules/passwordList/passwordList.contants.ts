export enum variantNames {
    connectCollapsed = 'connectCollapsed',
    connectExpanded = 'connectExpanded',
    entityFormCollapsed = 'entityFormCollapsed',
    entityFormExpanded = 'entityFormExpanded',
}

export type VariantName = keyof typeof variantNames;

export const placeholderEntityValue = 'XXX';

export interface PasswordEntityVulnerablePayload {
    password: string;
    login: string;
}

export interface PasswordEntityPayload extends PasswordEntityVulnerablePayload {
    label: string;
}

export interface PasswordEntityPayloadReferable extends PasswordEntityPayload {
    refId: string;
}
