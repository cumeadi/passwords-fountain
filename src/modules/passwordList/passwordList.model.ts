import { AppState } from '@/store';
import { mergeState } from '@/common/utils/store';

const change = mergeState('passwordList');

export const passwordListState = {} as const;

export type PasswordListState = typeof passwordListState;

export const passwordListActions = {
    noop(appState: AppState): Partial<AppState> {
        return change({});
    },
} as const;

export { useAction } from '@preact-hooks/unistore';
