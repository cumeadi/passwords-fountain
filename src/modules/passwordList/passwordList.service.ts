import { decrypt } from '@/modules/cipher/cipher.service';
import {
    PasswordEntity,
    PasswordEntityPayload,
    PasswordEntityRaw,
} from '@/modules/database/database.service';

// TODO: check if really needed
const normalizePasswordEntities = (
    rawPasswordList: PasswordEntityRaw[],
    masterKey: string
): PasswordEntity[] => {
    return rawPasswordList.map(
        (entity: PasswordEntityRaw): PasswordEntity => {
            const decryptedEntity = decrypt(
                entity.data.value,
                masterKey,
                true
            ) as PasswordEntityPayload;

            return {
                refId: entity.ref.id,
                createdAt: String(entity.ts),
                label: decryptedEntity.label,
                password: decryptedEntity.password,
                login: decryptedEntity.login,
            };
        }
    );
};
