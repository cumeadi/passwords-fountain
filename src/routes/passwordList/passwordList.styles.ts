import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { media, stylelint, theme, typography } from '@/common/theme';
setPragma(h);

export const PasswordEntityWrapper = styled('div')`
    margin-bottom: ${theme.spacing.s12};
`;

export const Wrapper = styled('section')`
    padding-top: 100px;

    ${PasswordEntityWrapper}:last-of-type {
        margin-bottom: 0;
    }
`;

export const OptionsPanelWrapper = styled('div')`
    display: grid;
    justify-items: center;
    margin-bottom: ${theme.spacing.s12};
`;

export const Placeholder = styled('div')`
    display: grid;
    place-items: center;
    padding: 120px 0 0 0;

    ${media.gte(theme.breakpoints.s480)(stylelint.css`
        padding: 170px 0 0 0;
    `)}
`;

export const PlaceholderTextWrapper = styled('div')`
    ${typography.text24}
    text-align: center;
    color: ${theme.colors.darkBlue};
    padding: ${theme.spacing.l24} 0;
`;

export const IconSizer = styled('div')`
    margin: 0 auto;

    img {
        width: 150px;
        height: 150px;
    }

    ${media.gte(theme.breakpoints.s480)(stylelint.css`
        img {
            width: 200px;
            height: 200px;
        }
    `)}
`;
