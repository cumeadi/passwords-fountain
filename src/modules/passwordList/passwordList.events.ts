import mitt from 'mitt';
import { asyncListener, syncListener } from '@/common/utils/eventListeners';
import { store } from '@/store';
import { passwordListActions } from './passwordList.model';
const emitter = mitt();

export enum passwordListEventTypes {
    noop = 'noop',
}

emitter.on(
    passwordListEventTypes.noop,
    asyncListener(() => {
        console.log(
            "When emitted from reducer's handler - will be called after state change."
        );
    })
);

emitter.on(
    passwordListEventTypes.noop,
    syncListener(() => {
        console.log(
            "When emitted from reducer's handler - will be called before state change."
        );
    })
);

export { emitter as passwordListEmitter };
