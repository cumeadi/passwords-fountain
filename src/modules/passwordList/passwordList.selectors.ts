import { createSelector } from 'reselect';
import { AppState } from '@/store';
import { PasswordListState } from './passwordList.state';
import { PasswordEntityPayloadReferable } from '@/modules/passwordList/passwordList.constants';

const selectPasswordList = (state: AppState): PasswordListState =>
    state.passwordList;

const mock = [
    {
        ref: { id: 'x1', value: { id: 'x1 ' } },
        ts: Date.now(),
        data: { label: 'Bank account', value: 'v1' },
    },
    {
        ref: { id: 'x2', value: { id: 'x2 ' } },
        ts: Date.now(),
        data: { label: 'Origin', value: 'v2' },
    },
    {
        ref: { id: 'x3', value: { id: 'x3 ' } },
        ts: Date.now(),
        data: { label: 'Steam', value: 'v3' },
    },
    {
        ref: { id: 'x4', value: { id: 'x4' } },
        ts: Date.now(),
        data: { label: 'Xiaomi', value: 'v4' },
    },
];

export const selectPasswords = createSelector(
    selectPasswordList,
    (passwordListState: PasswordListState) => passwordListState.passwords
);

export const selectCurrentOptionPanelVariantName = createSelector(
    selectPasswordList,
    (passwordListState: PasswordListState) =>
        passwordListState.currentOptionPanelVariantName
);

export const selectSelectedAndDecryptedEntity = createSelector(
    selectPasswordList,
    (passwordListState: PasswordListState) =>
        passwordListState.selectedAndDecryptedEntity
);

export const selectIsInEditMode = createSelector(
    selectSelectedAndDecryptedEntity,
    (selectedAndDecryptedEntity: PasswordEntityPayloadReferable) =>
        Object.keys(selectedAndDecryptedEntity).length !== 0
);

export const selectSelectedAndDecryptedEntityByRefId = (refId: string) =>
    createSelector(
        selectSelectedAndDecryptedEntity,
        (selectedAndDecryptedEntity: PasswordEntityPayloadReferable) => {
            if (selectedAndDecryptedEntity.refId === refId) {
                return selectedAndDecryptedEntity;
            }

            return {} as PasswordEntityPayloadReferable;
        }
    );
