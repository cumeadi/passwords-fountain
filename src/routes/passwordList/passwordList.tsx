import { h, VNode } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import { useSelector } from '@/store';
import {
    Wrapper,
    PasswordEntityWrapper,
    OptionsPanelWrapper,
    Placeholder,
    PlaceholderTextWrapper,
    IconSizer,
} from './passwordList.styles';
import {
    selectIsClientSet,
    selectIsFirstTimeOnDevice,
} from '@/modules/database/database.selectors';
import { route } from 'preact-router';
import { useEffect, useState } from 'preact/hooks';
import { OptionsPanel } from '@/modules/passwordList/components/optionsPanel';
import { selectPasswords } from '@/modules/passwordList/passwordList.selectors';
import { PasswordEntityRaw } from '@/modules/database/database.service';
import { PasswordEntity } from '@/modules/passwordList/components/passwordEntity';
import { useAction } from '@preact-hooks/unistore';
import { passwordListActions } from '@/modules/passwordList/passwordList.actions';
import { optionsPanelVariantNames } from '@/modules/passwordList/passwordList.constants';
import { renderIfTrue } from '@/common/utils/rendering';
import { Icon } from '@/common/components/icon';
import { Text } from '@/modules/localisation/components/text';

export const useFirstTimeRedirection = (): void => {
    useEffect(() => {
        const isFirstTimeOnDevice = useSelector(selectIsFirstTimeOnDevice);

        if (isFirstTimeOnDevice && !(window as any).prerendering) {
            // User has to be connected to database before accessing /app page
            setTimeout(() => {
                route('/settings');
            });
        }
    }, []);
};

export const PasswordList: TypedComponent<Props> = () => {
    useFirstTimeRedirection();
    const passwords = useSelector(selectPasswords);
    const isClientSet = useSelector(selectIsClientSet);
    const shouldRenderPlaceholder =
        !isClientSet || (isClientSet && !Boolean(passwords.length));
    const switchOptionsPanelVariant = useAction(
        passwordListActions.switchOptionPanelVariant
    );
    const [
        selectedEntity,
        setSelectedEntity,
    ] = useState<PasswordEntityRaw | null>(null);

    useEffect(() => {
        if (isClientSet) {
            switchOptionsPanelVariant(
                optionsPanelVariantNames.entityFormCollapsed
            );
        }
    }, []);

    const renderPasswords = (): VNode[] => {
        return passwords.map(
            (entity: PasswordEntityRaw): VNode => (
                <PasswordEntityWrapper key={entity.ref.id}>
                    <PasswordEntity
                        data={entity}
                        isSelected={selectedEntity === entity}
                        onClick={setSelectedEntity}
                    />
                </PasswordEntityWrapper>
            )
        );
    };

    const renderPlaceholder = renderIfTrue(
        (): VNode => {
            const iconName = isClientSet ? 'database' : 'padlock';
            const text = isClientSet
                ? 'passwordList.noDataPlaceholder'
                : 'passwordList.connectionPlaceholder';
            return (
                <Placeholder>
                    <IconSizer>
                        <Icon name={iconName} />
                    </IconSizer>
                    <PlaceholderTextWrapper>
                        <Text>{text}</Text>
                    </PlaceholderTextWrapper>
                </Placeholder>
            );
        }
    );

    return (
        <Wrapper>
            <OptionsPanelWrapper>
                <OptionsPanel />
            </OptionsPanelWrapper>
            {renderPasswords()}
            {renderPlaceholder(shouldRenderPlaceholder)}
        </Wrapper>
    );
};

export interface Props {}
