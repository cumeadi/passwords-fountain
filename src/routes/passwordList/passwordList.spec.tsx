import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { PasswordList } from './passwordList';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {
    overlay: {
        isGlobalLoaderVisible: false,
    },
    database: {
        client: {},
    },
    passwordList: {
        passwords: [],
    },
};

describe('PasswordList', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <StoreProvider value={createStore(stateMock)}>
                <PasswordList />
            </StoreProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
