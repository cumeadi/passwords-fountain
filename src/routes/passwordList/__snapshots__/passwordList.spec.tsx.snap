// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordList should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go103570428"
  >
    <div
      class="go1530330403"
    >
      <section
        class="go1009640918"
      >
        <undefined />
      </section>
    </div>
    <ul />
    <div
      class="go482900832"
    >
      <div
        class="go1566361126"
      >
        <img
          alt="icon"
          class="go115679"
          height="100"
          src="static-asset-stub"
          width="100"
        />
      </div>
      <div
        class="go710061235"
      >
        Connect to database to see your passwords here
      </div>
    </div>
  </section>
</DocumentFragment>
`;
